File: fsharp-2017/src/Day04.fs
module Day04

open Util
open System

let hasDup(a:string[]) : bool =
  let mutable seen = set []
  a |> Array.exists (fun s ->
    if not (seen.Contains s) then seen <- seen.Add s; false
    else true
  )

let hasAnagram(a:string[]) : bool =
  let mutable seen = set []
  a |> Array.exists (fun s ->
    let hash = s |> Seq.sort |> String.Concat
    if not (seen.Contains hash) then seen <- seen.Add hash; false
    else true
  )

let countValids f aa = 
  aa |> Array.filter f |> Array.length

let run (instr:string) =
  let aa = instr |> splitLineAnd ' '
  printfn "Day 4-1 answer: %A" <| countValids (not << hasDup) aa
  printfn "Day 4-2 answer: %A" <| countValids (not << hasAnagram) aa

File: js-pipe-fkit_2017/src/day04.js
const { filter, sort, id } = require('fkit')
const { strToWords2D } = require('./helpers')

const noDups = a => notInAry(id, a)
const noAnagrams = a => notInAry(sort, a)

const notInAry = (f, strs) => {
  let seen = {}
  for (let s of strs) {
    const hash = f(s)
    if (!seen[hash]) seen[hash] = true;
    else return false
  }
  return true
}

const countValids = (f, aa) => (aa |> filter(f)).length

const run = (inStr, log) => {
  const aa = strToWords2D(inStr)
  log.p1( countValids(noDups, aa))  // 466
  log.p2( countValids(noAnagrams, aa)) // 251
}
module.exports = {
  run,
  countValids,
  noDups,
  noAnagrams,
}